### PR TITLE
[Messenger] dispatch event when a message is retried

### PR DESCRIPTION
**Pull request by nikophil**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT

Hello,

i'm working on a bundle which helps to monitor messenger queues (add some stats for queues/transports + ability to manage failed messages from the browser)
https://github.com/SymfonyCasts/messenger-monitor-bundle/

and we're missing some hooks in the messaging system:
1. a way to know when a message has been retried (fixed by dispatching a new `WorkerMessageRetriedEvent` in `SendFailedMessageForRetryListener::onMessageFailed()`)
2. a way to update the enveloppe in worker message events (fixed by adding `AbstractWorkerMessageEvent::setEnvelope()`)

if needed i can provide some precise use cases.

thanks.

### CODE REVIEW

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php
Comment (weaverryan) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php]: We're missing the service wiring to actually inject the new argument

Comment (nikophil) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php]: indeed

**Comment by chalasr** [comment]
> a way to update the enveloppe in worker message events (fixed by adding AbstractWorkerMessageEvent::setEnvelope())

Can you please provide a use case for this?

**Comment by nikophil** [comment]
@chalasr we're aggregating some metadata for each message (`received_at`, `waiting_time`, etc...)
thus, we in order to track messages we affect an [id with a stamp once message is sent to transport](https://github.com/SymfonyCasts/messenger-monitor-bundle/blob/master/src/EventListener/AddStampOnMessageSentListener.php) leveraging `SendMessageToTransportsEvent::setEnvelope()`.

But if a message is sent by another system (eg: in a micro-services architecture), we cannot use `SendMessageToTransportsEvent` and need another hook to add this id.

**Comment by chalasr** [comment]
@nikophil Works for me, thanks.

Can you add some CHANGELOG entries?

File: src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php
Comment (Tobion) [inline src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php]: making the envelope mutable means, all the dispatching of events would need to be changed to refetch the potentially modified envelope. -1 on doing this.

**Review by Tobion — CHANGES_REQUESTED** [review]
-1 on this feature as people can just use the WorkerMessageFailedEvent with a priority that comes after the SendFailedMessageForRetryListener

File: src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php
Comment (weaverryan) [reply src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php]: Yes, this part is missing. But do you see it as a huge problem to (effectively) add `$envelope = $event->getEnvelope()`? I made the events immutable originally because I feel that it's always better to be conservative *until* we have a use-case. But this PR now does have a use-case.

Comment (nikophil) [reply src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php]: sorry i missed that...
in case this PR is accepted, i've fixed that and refetched envelope from mutable events.
i've added two new tests which check that events mutability is handled

Comment (weaverryan) [reply src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php]: Did we need to add this in the abstract class? This will affect (on master currently) 3 classes. Maybe we need hook points in all of them? But in this PR, we only need a hook point in one of them.

Comment (nikophil) [reply src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php]: ok, i'll do that, it will be less error prone this way

Comment (nikophil) [reply src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php]: done :)

**Comment by weaverryan** [comment]
I'm still supportive of this feature, but we would need @Tobion to reconsider his vote :)

**Comment by fabpot** [comment]
@Tobion Can you have another look at this PR?

File: src/Symfony/Component/Messenger/Tests/WorkerTest.php
Comment (Tobion) [inline src/Symfony/Component/Messenger/Tests/WorkerTest.php:259]: wrong type hint. should be `WorkerMessageReceivedEvent`

File: src/Symfony/Component/Messenger/Tests/WorkerTest.php
Comment (Tobion) [inline src/Symfony/Component/Messenger/Tests/WorkerTest.php:278]: Language: `acknowledgedEnvelopes`

File: src/Symfony/Component/Messenger/CHANGELOG.md
Comment (Tobion) [inline src/Symfony/Component/Messenger/CHANGELOG.md:11]: wrong changelog: `WorkerMessageReceivedEvent::setEnvelope()`

File: src/Symfony/Component/Messenger/Event/WorkerMessageReceivedEvent.php
Comment (Tobion) [inline src/Symfony/Component/Messenger/Event/WorkerMessageReceivedEvent.php:34]: missing void return

File: src/Symfony/Component/Messenger/Event/WorkerMessageRetriedEvent.php
Comment (Tobion) [inline src/Symfony/Component/Messenger/Event/WorkerMessageRetriedEvent.php:15]: Dispatched after a message has been sent for retry.

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php
Comment (Tobion) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php:76]: please inverse the logic here `if (null !== $this->eventDispatcher) {`. no point in early return here.

**Review by Tobion — CHANGES_REQUESTED** [review]

**Comment by fabpot** [comment]
@nikophil Still willing to finish this PR?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:106]: Must be moved to the new PHP config file (the XML file does not exist anymore in current master).

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by nikophil** [comment]
Hi, yes sorry, i'll do that shortly

**Comment by nikophil** [comment]
@fabpot @Tobion i've fixed the PR, sorry for the delay

File: src/Symfony/Component/Messenger/CHANGELOG.md
Comment (weaverryan) [inline src/Symfony/Component/Messenger/CHANGELOG.md:23]: I believe this section should be removed

File: src/Symfony/Component/Messenger/Event/WorkerMessageReceivedEvent.php
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Event/WorkerMessageReceivedEvent.php:37]: In #32904, we have this same need and we're doing it with an `addStamps()` method on `AbstractWorkerMessageEvent`. I believe we should use that approach. If you agree, it means we should wait for that PR to be merged or (better) duplicate that functionality here exactly so that either PR can be merged first without conflicts (if the code is duplicated *exactly*).

File: src/Symfony/Component/Messenger/Worker.php
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Worker.php:109]: If we do the `addStamps()` idea, then this can be removed

**Review by weaverryan — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/Worker.php
Comment (weaverryan) [reply src/Symfony/Component/Messenger/Worker.php:109]: I'm wrong about this - this needs to stay

**Review by weaverryan — APPROVED** [review]
This is ready - it duplicates some logic from #32904, as both PR's needed the same extension point.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nikophil.

**Pull request merged**
